src/main/java/com/epam/esm/utils/QueryConstructor.java

```diff
@@ -0,0 +1,77 @@
+package com.epam.esm.utils;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.SqlQueries;
+import org.springframework.stereotype.Component;
+
+@Component
+public class QueryConstructor {
+
+    public String constructQuery(String tagName, String giftCertificateName, String description, String sortByName, String sortByDate) {
+        StringBuilder query = new StringBuilder(SqlQueries.FIND_ALL_CERTIFICATES);
+
+        if (giftCertificateName != null) {
+            query.append(" WHERE gc.name LIKE '%").append(giftCertificateName).append("%'");
+
+            if (description != null) {
+                query.append(" AND ");
+            }
+        }
+
+        if (description != null) {
+            if (giftCertificateName == null) {
+                query.append(" WHERE");
+            }
+
+            query.append(" gc.description LIKE '%").append(description).append("%'");
+        }
+
+        if (tagName != null) {
+            query.append(" HAVING gc.id IN (SELECT gc.id FROM gift_certificate gc JOIN gift_certificate_tag gct ON gct.gift_certificate_id = gc.id JOIN tag t ON gct.tag_id = t.id WHERE t.name = '").append(tagName).append("')");
```

Comment:
Please move this query `(SELECT gc.id FROM gift_certificate ...` somewhere

Reply:
Moved to the SqlQueries class